### PR TITLE
AG-Grid

### PR DESCRIPTION
**Pull request by Gustav-Eikaas**
# Description

AG-Grid package pretty much does everything we need. The only thing that is hard is persisting the configuration of the current view. Ag grid makes a controller/API on mount when the onGridReady callback is called. However if you store the grid's API in a variable and reference it after the grid has unmounted you get a warning saying that the grid has been destroyed because the view unmounted. The hard part is writing a controller that mirrors the grid's API in a good way. Preventing the grid's API from destroying itself seems like a futile effort towards a hacky solution.

For now most of the important state like rowData, column definitions etc are stored in a controller. But down the line we will probably face some issues with this. Specifically regarding bookmarks/viewSettings/localstorage etc..

Completes: AB#FILL_IN_YOUR_ISSUE_ID

## Checklist

- [x] I have performed a self-review of my own code.
- [ ] I have linked my DevOps task using the AB# tag.
- [x] My code is easy to read.
- [x] Documentation and comments is added where needed.
- [x] My code is covered by tests.

### CODE REVIEW

File: packages/ag-grid/src/lib/functions/registerCallback.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+export function registerCallback<TCallback>(
+    cb: TCallback,
+    list: Callback<TCallback>[],
+    unsub: (id: string) => void
+): OnCallbackSet {
+    const id = generateUniqueId();
+    list.push({ callback: cb, id });
+    return {
+        id,
+        unSubscribe: () => unsub(id),
```
Comment (Gustav-Eikaas) [inline packages/ag-grid/src/lib/functions/registerCallback.ts:12]: unsubscribe

File: packages/ag-grid/src/lib/test/onGridOptionsChanged.spec.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import { GridController } from '../classes';
+
+const mocFunction = jest.fn();
+
+const gridController = new GridController();
+
+/**
+ * Test to make sure the onGridOptionsChanged callbacks are called and passing the correct data
+ */
+describe('onGridOptionsChanged', () => {
```
Comment (Gustav-Eikaas) [inline packages/ag-grid/src/lib/test/onGridOptionsChanged.spec.ts:10]: Describe

File: packages/ag-grid/src/lib/test/onRowDataChanged.spec.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import { GridController } from '../classes';
+
+const mocFunction = jest.fn();
+
+const gridController = new GridController();
+
+/**
+ * Test to make sure the onRowDataChanged callbacks are called and passing the correct data
+ */
+describe('OnRowDataChangedCallback', () => {
```
Comment (Gustav-Eikaas) [inline packages/ag-grid/src/lib/test/onRowDataChanged.spec.ts:10]: Describe

File: packages/ag-grid/README.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# ag-grid
+
+This library was generated with [Nx](https://nx.dev).
```
Comment (Gustav-Eikaas) [inline packages/ag-grid/README.md:3]: Add readme

File: packages/ag-grid/src/lib/classes/gridController.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+     */
+    setGridOptions = (gridOptions: GridOptions) => {
+        this.gridOptions = gridOptions;
+        this.onGridOptionsChangedCallbacks.forEach(({callback}) => callback(gridOptions, this))
+    };
+
+    /**
+     * Registers a function to be called upon when grid options changes
+     * @param cb 
+     * @returns
```
Comment (Gustav-Eikaas) [inline packages/ag-grid/src/lib/classes/gridController.ts:34]: tsdoc

File: packages/ag-grid/src/lib/classes/gridController.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+import {
+    Callback,
+    OnCallbackSet,
+    OnGridOptionsChangedCallback,
+    OnRowDataChangedCallback,
+} from '../types';
+
+
+export class GridController<T> {
+    rowData: T[] = [];
```
Comment (Gustav-Eikaas) [inline packages/ag-grid/src/lib/classes/gridController.ts:12]: Tsdoc

File: packages/ag-grid/src/lib/classes/gridController.ts [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+     */
+    setRowData = (newData: T[]) => {
+        this.rowData = newData;
+        this.onRowDataChangedCallbacks.forEach(({ callback }) => callback(newData, this));
+    };
+
+    /**
+     * Registers a function to be called upon when row data changes
+     * @param callback 
+     * @returns
```
Comment (Gustav-Eikaas) [inline packages/ag-grid/src/lib/classes/gridController.ts:57]: tsdoc

File: packages/ag-grid/src/lib/classes/gridController.ts [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+    private unsubOnGridOptionsChanged = (id: string) => {
+        this.onGridOptionsChangedCallbacks = this.onGridOptionsChangedCallbacks.filter(
+            (s) => s.id !== id
+        );
+    };
+
+    /**
+     * Sets new row data and triggers all the onRowDataChanged callback's.
+     * @param newData
```
Comment (Gustav-Eikaas) [inline packages/ag-grid/src/lib/classes/gridController.ts:47]: tsdoc

File: packages/ag-grid/src/lib/hooks/useRowData.ts [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import { useEffect, useState } from "react";
+import { GridController } from "../classes";
+
+export function useRowData<TData>(controller: GridController<TData>){
+    const [rowData, setRowData] = useState(controller.rowData);
+
+    useEffect(() => {
+        const {unsubscribe: unSubscribe} = controller.onRowDataChanged(setRowData);
```
Comment (Gustav-Eikaas) [inline packages/ag-grid/src/lib/hooks/useRowData.ts:8]: unsub

File: packages/ag-grid/src/lib/test/onRowDataChanged.spec.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
+import { GridController } from '../classes';
+
+const mocFunction = jest.fn();
+
+const gridController = new GridController();
+
+/**
+ * Test to make sure the onRowDataChanged callbacks are called and passing the correct data
+ */
+describe('Tests for validating that callbacks are called when row data changes works', () => {
```
Comment (Gustav-Eikaas) [inline packages/ag-grid/src/lib/test/onRowDataChanged.spec.ts:10]: works

File: packages/ag-grid/src/lib/test/registerCallback.spec.ts [hunk 11]
```diff
@@ -0,0 +1,11 @@
+import { GridController } from "../classes"
+
+describe("Should add callback to array", () => {
+    it("Callback with corresponding id should be present", () => {
+        const controller = new GridController();
+        const {id, unSubscribe} = controller.onRowDataChanged(s => s)
```
Comment (Gustav-Eikaas) [inline packages/ag-grid/src/lib/test/registerCallback.spec.ts:6]: unsubscribe

**Review by jaysencpp — APPROVED** [review]

**Pull request merged**